torch/fx/graph_module.py

```diff
@@ -0,0 +1,459 @@
+linecache.getlines = patched_getline
+
+
+
+
+# this is fixed on master, WAR for 1.5
+def _find_module_of_method(orig_method):
+    name = orig_method.__name__
+    module = orig_method.__module__
+    if module is not None:
+        return module
+    for guess in [torch, torch.nn.functional]:
+        if getattr(guess, name, None) is orig_method:
+            return guess.__name__
+    raise RuntimeError(f'cannot find module for {orig_method}')
+
+HAS_VARSTUFF = inspect.CO_VARARGS | inspect.CO_VARKEYWORDS
+def _patch_function(fn, nargs):
+    co = fn.__code__
+    co_flags = co.co_flags & ~HAS_VARSTUFF
+    if hasattr(co, "co_posonlyargcount"):
+        co_args = (
+            nargs, 0,
+            0, co.co_nlocals, co.co_stacksize,
+            co_flags, co.co_code, co.co_consts, co.co_names,
+            co.co_varnames, co.co_filename, co.co_name,
+            co.co_firstlineno, co.co_lnotab, co.co_freevars,
+            co.co_cellvars
+        )
+    else:
+        co_args = (
+            nargs, 0, co.co_nlocals,
+            co.co_stacksize, co_flags, co.co_code, co.co_consts,
+            co.co_names, co.co_varnames, co.co_filename,
+            co.co_name, co.co_firstlineno, co.co_lnotab,
+            co.co_freevars, co.co_cellvars)
+    new_code = CodeType(*co_args)
+    return FunctionType(new_code, fn.__globals__, fn.__name__, fn.__defaults__, fn.__closure__)
+
+    # we need to insert placeholder nodes for *args, and **kwargs,
+    # so we can't call this function normally, otherwise it would try to unpack them
+    # instead, let's make python think that args and kwargs are normay variables
+
+
+
+class TraceError(ValueError):
+    pass
+
+# Nodes represent a definition of a value in our graph of operators.
+# For simplicitly, they also have methods defined on them so that they serve as proxy objects for _building_
+# more nodes.
+class Node:
+    def __init__(self, graph, name, op, target, args, kwargs):
+        self.graph = graph
+        self.name = name  # unique name of value being created
+        self.op = op  # the kind of operation = placeholder|call_method|call_module|call_function|getattr
+        self.target = target  # for method/module/function, the name of the method/module/function/attr
+        # being invoked, e.g add, layer1, or torch.add
+        self.args = args
+        self.kwargs = kwargs
+        self.uses = 0
+
+    def __repr__(self):
+        return self.name
+
+    def __getattr__(self, k):
+        # note: not added to the graph yet, if this is a method call
+        # we peephole optimize to the method invocation
+        return Attribute(self, k)
+
+    def __call__(self, *args, **kwargs):
+        return self._create_node('call_method', '__call__', [self] + args, kwargs)
+
+    def __iter__(self):
+        frame = inspect.currentframe()
+        calling_frame = frame.f_back
+        inst = list(dis.get_instructions(calling_frame.f_code))[calling_frame.f_lasti // 2]
```

Comment:
yeah this code is super confusing, who wrote this!? (I wrote this...) Normally, unpacking a tuple would try to call `__iter__`. In most cases, we want to error on calling `__iter__` because the length of the symbolic value being iterated is unknown. However, in the pattern matching case `x, y = rhs` we know `rhs` must have length 2, otherwise this will error. This code snoops the callers code to see if an unpack is taking place and that there are two elements being matched. It then returns two things in the iterator to make sure the match succeeds. This makes things like `torch.maxi` which returns two values traceable when it otherwise wouldn't be. However, it isn't strictly needed since you can always rewrite the calling code to explicitly match `x = rhs[0]; y = rhs[1]`.